ouroboros-consensus/test-storage/Test/Ouroboros/Storage/ChainDB/VolDB.hs

```diff
@@ -0,0 +1,340 @@
+
+{-------------------------------------------------------------------------------
+  Top-level tests
+-------------------------------------------------------------------------------}
+
+tests :: TestTree
+tests = testGroup "VolDB" [
+      testProperty "isReachable" prop_isReachable
+    ]
+
+{-------------------------------------------------------------------------------
+  Test properties
+-------------------------------------------------------------------------------}
+
+prop_isReachable :: Pretty (ReachableSetup TestBlock) -> Property
+prop_isReachable (Pretty ReachableSetup { currentChain, forkTip, fork, blockInfo }) =
+        Pretty (isReachable (flip Map.lookup blockInfo) currentChain forkTip)
+    === Pretty fork
+
+{-------------------------------------------------------------------------------
+  Test setup
+-------------------------------------------------------------------------------}
+
+data ReachableSetup blk = ReachableSetup {
+      currentChain :: AnchoredFragment (Header blk)
+    , forkTip      :: RealPoint blk
+    , fork         :: Maybe (PointChainDiff blk)
+    , blockInfo    :: Map (HeaderHash blk) (VolDB.BlockInfo (HeaderHash blk))
+    }
+
+deriving instance (HasHeader blk, Show (Header blk)) => Show (ReachableSetup blk)
+
+-- | Newtype that uses 'ppShow' for its 'Show' instance.
+newtype Pretty a = Pretty a
+  deriving (Eq, Arbitrary)
+
+instance Show a => Show (Pretty a) where
+  show (Pretty a) = ppShow a
+
+{-------------------------------------------------------------------------------
+  Generators
+-------------------------------------------------------------------------------}
+
+-- | All blocks on the current chain use this value as for their body. This
+-- allows us to statically convert a header from the current chain to a block.
+currentChainBody :: TestBody
+currentChainBody = TestBody 0 True
+
+genFirstBlock ::
+     Gen TestBody
+  -> Gen TestBlock
+genFirstBlock genBody = oneof
+    [ firstBlock 0            <$> genBody
+    , firstEBB   (const True) <$> genBody
+    ]
+
+genSuccessor ::
+     TestBlock
+  -> Gen TestBody
+  -> Gen TestBlock
+genSuccessor prevBlk genBody = frequency
+    [ ( if prevIsEBB then 0 else 1
+      , mkNextEBB (const True) prevBlk
+          <$> chooseSlot (prevSlot + 1) (prevSlot + 2)
+          -- We don't care about EpochNo
+          <*> return 0
+          <*> genBody
+      )
+    , (3
+      , mkNextBlock prevBlk
+          <$> chooseSlot (prevSlot + if prevIsEBB then 0 else 1) (prevSlot + 2)
+          <*> genBody
+      )
+    ]
+  where
+    prevSlot  = blockSlot prevBlk
+    prevIsEBB = isJust $ headerIsEBB $ getHeader prevBlk
+
+    chooseSlot :: SlotNo -> SlotNo -> Gen SlotNo
+    chooseSlot (SlotNo start) (SlotNo end) = SlotNo <$> choose (start, end)
+
+genChainHelper ::
+     Gen TestBody
+  -> Int                                  -- ^ Number of headers to generate
+  -> TestBlock                            -- ^ Previous block
```

Comment:
Take `WithOrigin TestBlock` (or `WithOrigin HeaderFields`) and do the choice between `genFirstBlock`/`genSuccessor` here.